### PR TITLE
Fix #874 "Singleplayer freeoriond opens socket with INADDR_ANY triggering firewalls"

### LINKED ISSUE #874: Singleplayer freeoriond opens socket with INADDR_ANY triggering firewalls
### Environment

- **FreeOrion Version**: master
- **Operating System**: any
- **Graphic card used**: -
- **Fetched as**
  - Compiled from source
### Description

As reported with #872 `freeoriond` seems to trigger various Desktop Firewalls to ask for permission when opening a network connection, even when the user only plays a single player game.  Unaware users may block this (why should a single player game require 'internet access' anyway?) and are unable to play any single player game unless the application is removed from the firewall blacklist.
### Expected Result

User is never pestered with any Firewall permission query when starting a single player game.
#### Steps to reproduce

Prerequisites:
- Have a running firewall without a freeoriond whitelist entry.

Steps:
- Run freeorion.
- Do a `Quick Start` game.
- Firewall asks user if `freeoriond` should be allowed to have a network connection.
- Deny network connection for `freeoriond`.
- User can't play any game now or in the future because `freeoriond` is denied to create a network connection.

### PR DESCRIPTION
**Pull request by adrianbroher**
This pull request implements various features required to fix #874.

A review would be appreciated.

### CODE REVIEW

File: default/stringtables/en.txt [hunk 1]
```diff
@@ -1113,6 +1113,9 @@ COMMAND_LINE_DEFAULT
 OPTIONS_DB_HELP
 Print this help message.
 
+OPTIONS_DB_SINGLEPLAYER
+Start server in single player host mode.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: Could you elaborate on what that means?

File: client/human/HumanClientApp.cpp [hunk 2]
```diff
@@ -409,6 +409,10 @@ void HumanClientApp::StartServer() {
     } else {
         DebugLogger() << "ai-config not set.";
     }
+    if (m_single_player_game)
+    {
```
Comment (geoffthemedio) [inline client/human/HumanClientApp.cpp]: { to previous line please...

File: network/ServerNetworking.cpp [hunk 3]
```diff
@@ -309,12 +310,20 @@ ServerNetworking::ServerNetworking(boost::asio::io_service& io_service,
...
-    m_discovery_server(new DiscoveryServer(io_service)),
+    m_discovery_server(0),
     m_player_connection_acceptor(io_service),
     m_nonplayer_message_callback(nonplayer_message_callback),
     m_player_message_callback(player_message_callback),
     m_disconnected_callback(disconnected_callback)
-{ Init(); }
+{
+    if (!GetOptionsDB().Get<bool>("singleplayer"))
+    {
```
Comment (geoffthemedio) [inline network/ServerNetworking.cpp]: { to prev line

File: network/ServerNetworking.cpp [hunk 4]
```diff
@@ -475,6 +484,12 @@ void ServerNetworking::SetHostPlayerID(int host_player_id)
 
 void ServerNetworking::Init() {
     tcp::endpoint endpoint(tcp::v4(), MESSAGE_PORT);
+
+    if(GetOptionsDB().Get<bool>("singleplayer"))
```
Comment (geoffthemedio) [inline network/ServerNetworking.cpp]: space after if, { to this line

File: network/ServerNetworking.cpp [hunk 5]
```diff
@@ -475,6 +484,12 @@ void ServerNetworking::SetHostPlayerID(int host_player_id)
...
+    {
+        endpoint.address(boost::asio::ip::address_v4::loopback());
```
Comment (geoffthemedio) [inline network/ServerNetworking.cpp:493]: can you add a brief comment here about what this does / why?

File: default/stringtables/en.txt (hunk 1 above)
Comment (adrianbroher) [inline default/stringtables/en.txt]: Here, as comment or in the translation string?

Comment (geoffthemedio) [inline default/stringtables/en.txt]: In the string

**Comment by adrianbroher** [comment]
Changed code according to code review feedback.

**Comment by adrianbroher** [comment]
Also some Windows user should check if the problem with the Desktop firewalls has been fixed with this PR.

**Comment by geoffthemedio** [comment]
still works ok for single player games for me

**Comment by adrianbroher** [comment]
Did you remove freeorion from the firewall whitelist?
Did no firewall notification show up for freeorion when starting a single player game?

**Comment by geoffthemedio** [comment]
freeoriond.exe was removed from the whitelist. After that, starting a single player game didn't produce a windows firewall popup, but starting a multiplayer game did.

**Comment by adrianbroher** [comment]
That's exactly the behavior I want to see.

Are there any objections left that prevent merging? Is the updated translation string sufficient?

**Comment by geoffthemedio** [comment]
no; seems fine

**Comment by Vezzra** [comment]
Just did some quick tests on OSX and can confirm that this patch works exactly as intended on that platform too.

Single player games don't trigger a firewall popup, multiplayer games do. Without this patch the firewall popup gets triggered for both.

**Pull request merged**